pootle/apps/virtualfolder/utils.py

```diff
@@ -0,0 +1,233 @@
+        self.vf = vf
+
+    @property
+    def existing_stores(self):
+        """Currently associated Stores"""
+        return self.vf.stores.all()
+
+    @property
+    def languages(self):
+        """The languages associated with this vfolder
+        If `all_languages` is set then `None` is returned
+        """
+        if self.vf.all_languages:
+            return None
+        return self.vf.languages.all()
+
+    @property
+    def projects(self):
+        """The projects associated with this vfolder
+        If `all_projects` is set then `None` is returned
+        """
+        if self.vf.all_projects:
+            return None
+        return self.vf.projects.all()
+
+    @property
+    def matching_stores(self):
+        """Store qs containing all stores that match
+        project, language, and rules for this vfolder
+        """
+        return self.filter_from_rules(self.store_qs)
+
+    @property
+    def rules(self):
+        """Glob matching rules"""
+        return (
+            "%s" % r.strip()
+            for r
+            in self.vf.filter_rules.split(","))
+
+    @property
+    def store_manager(self):
+        """The root object manager for finding/adding stores"""
+        return Store.objects
+
+    @property
+    def store_qs(self):
+        """The stores qs without any rule filtering"""
+        return self.filter_projects(
+            self.filter_languages(
+                self.store_manager))
+
+    def update_stores(self):
+        """Add and delete Store associations as necessary, and set the
+        priority for any affected Stores
+        """
+        added, removed = self.add_and_remove_stores()
+        for store in added:
+            update_vfolder_tree(self.vf, store)
+            if store.priority != self.vf.priority:
+                store.set_priority()
+        for store in removed:
+            if store.priority == self.vf.priority:
+                store.set_priority()
+
+    def add_and_remove_stores(self):
+        """Add Stores that should be associated but arent, delete Store
+        associations for Stores that are associated but shouldnt be
+        """
+        existing_stores = set(self.existing_stores)
+        matching_stores = set(self.matching_stores)
+        to_add = matching_stores - existing_stores
+        to_remove = existing_stores - matching_stores
+        if to_add:
+            self.add_stores(to_add)
+        if to_remove:
+            self.remove_stores(to_remove)
+        return to_add, to_remove
+
+    def add_stores(self, stores):
+        """Associate a Store"""
+        self.vf.stores.add(*stores)
+
+    def remove_stores(self, stores):
+        self.vf.stores.remove(*stores)
+
+    def should_add_store(self, store):
+        if self.store_matches(store) and not self.store_associated(store):
+            return True
+        return False
+
+    def store_associated(self, store):
+        return self.vf.stores.through.objects.filter(store_id=store.id).exists()
```

Comment:
i checked this and this is the difference in sql

with this:
```
'SELECT (1) AS "a" FROM "virtualfolder_virtualfolder_stores" WHERE ("virtualfolder_virtualfolder_stores"."store_id" = %s AND "virtualfolder_virtualfolder_stores"."virtualfolder_id" = %s) LIMIT 1' - PARAMS = (31, 6)
```
and using `self.vf.stores.filter(store_id=store_id).exists()`

```
u'SELECT (1) AS "a" FROM "pootle_store_store" INNER JOIN "virtualfolder_virtualfolder_stores" ON ( "pootle_store_store"."id" = "virtualfolder_virtualfolder_stores"."store_id" ) WHERE ("virtualfolder_virtualfolder_stores"."virtualfolder_id" = %s AND "pootle_store_store"."id" = %s) LIMIT 1' - PARAMS = (6, 31)

```